### PR TITLE
expression: improve the compatibility with mysql when datatime is invalid.

### PR DESCRIPTION
**Pull request by ono-yoko**
### What is changed and how it works?
It is rather easy fix, just preparing a correct string and pass it into ErrorDataTimeFormat error. 
In this way, the wanting message looks correct.  

This PR closes https://github.com/pingcap/tidb/issues/11304

### Check List 

Tests 
 - Integration test

Code changes

Side effects
No side effects

Related changes
 - Need to cherry-pick to the release branch

### CODE REVIEW

**Comment by ono-yoko** [comment]
@XuHuaiyu @SunRunAway PTAL 

Thanks in advance. I do not have any right to run test or adding label. ;)

File: types/time.go [hunk 1]
```diff
@@ -1468,8 +1468,9 @@ func checkDateRange(t MysqlTime) error {
 }
 
 func checkMonthDay(year, month, day int, allowInvalidDate bool) error {
+	dateTimeInStr := fmt.Sprintf("%d-%d-%d", year, month, day)
```
Comment (SunRunAway) [inline types/time.go]: We do not need this line if there's no error, you can put `fmt.Sprintf` directly into `GenWithStackByArgs`.

Comment (ono-yoko) [reply types/time.go]: Nice point. Done.

**Comment by SunRunAway** [comment]
@ono-yoko Please take a look at the CI problem.

**Comment by ono-yoko** [comment]
@SunRunAway Done. It seems I forgot to check err in test.

**Comment by ono-yoko** [comment]
@SunRunAway I just checked ci failure. I do not think it has something to do with my change.

**Pull request merged**